### PR TITLE
Provide documentation for `Default` impls

### PR DESCRIPTION
**Pull request by frewsxcv**
For the most part, the `Default` impls are pretty obvious as far as what
we expect them to return, but I think it's better to explicitly state
what's going to be returned in the documentation instead of relying on
the user to make assumptions/look at the source.

### CODE REVIEW

File: src/libstd/collections/hash/state.rs [hunk 1]
```diff
@@ -51,5 +51,6 @@ impl<H> Clone for DefaultState<H> {
 }
 
 impl<H> Default for DefaultState<H> {
+    /// Creates a new `DefaultState` containing `marker::PhantomData`
```
Comment (sfackler) [inline src/libstd/collections/hash/state.rs]: `PhantomData` is an implementation detail of `DefaultState`. It shouldn't be documented.

File: src/libcore/default.rs [hunk 2]
```diff
@@ -137,6 +137,7 @@ pub trait Default {
 
 macro_rules! default_impl {
     ($t:ty, $v:expr) => {
+        /// Creates a new `$t` with value `$v`
```
Comment (sfackler) [inline src/libcore/default.rs]: Does this work?

Comment (frewsxcv) [inline src/libcore/default.rs]: I don't think so now that I think about it. I just removed it in the latest force push. I'm not going to worry about it for now

**Comment by frewsxcv** [comment]
@sfackler The latest force push look okay?

**Comment by steveklabnik** [comment]
>  I think it's better to explicitly state
> what's going to be returned in the documentation instead of relying on
> the user to make assumptions/look at the source.

In general, as you mention, we _only_ document impls when it does something different than you'd expect. Given this rule, I'm not sure that this patch works.

I will say that how Rustdoc should show trait stuff is very much under discussion, that's just the current way we handle things.

**Comment by frewsxcv** [comment]
> we only document impls when it does something different than you'd expect

@steveklabnik Just looking at the documentation, could you tell me what you'd 'expect' for the `Default` `impl` for `HashSet`? hint: it's not `HashSet::new`.

I don't see the harm in documenting what a returns.

**Comment by steveklabnik** [comment]
Like most containers, I would expect it to consider the defaults of what it's generic over, which is exactly what it does.

> I don't see the harm in documenting what a function returns.

I think the broader issue here isn't that more documentation is worse, it's just that we're unsure how to best do this documentation. And in my pull request for `Arc` docs, I actually did implement `Default`, but that didn't pass review due to the rule.

It's not that I think documentation is bad, it's just that consistency is also important. Maybe we should have a discussion about this topic in general.

**Comment by steveklabnik** [comment]
Since the current convention is not to do this, I'm going to give it a close. @frewsxcv , if you want to have a discussion, we could have one over in the RFCs repo about the pros and cons of changing the convention.

**Comment by frewsxcv** [comment]
> changing the convention.

@steveklabnik I'm not opposed this being closed, but for clarification, what convention are we talking about? Documenting `Default` `impl` or documenting `impl` in general?

**Comment by steveklabnik** [comment]
The current convention is:

> When writing an `impl`, only add documentation if that implementation does something non-obvious.

An example is `Clone` for `Arc<T>`: it doesn't actually copy any data, it bumps a reference count. So it gets documented. But `Clone` for `Vec<T>` doesn't, as it's just a deep copy.

**Comment by matthew-piziak** [comment]
Part of #36265.

**Pull request closed without merging**